scripts/linters/pylint_extensions.py

```diff
@@ -1654,99 +1654,110 @@ class SingleLineCommentChecker(checkers.BaseChecker):
+                # Comments must start with a space.
+                space_at_beginning_of_comment = (
+                    self._check_space_at_beginning_of_comment(line, line_num))
+
+                if line_num - previous_line_num > 1:
+                    prev_line_list = previous_line.split()
+
+                    if len(previous_line) > 1:
+                        excluded_phrase_is_present_at_end = any(
+                            word in prev_line_list[-1] for word in
+                            EXCLUDED_PHRASES)
+                        excluded_phrase_at_beginning_of_prev_line = any(
+                            word in prev_line_list[1] for word in
+                            EXCLUDED_PHRASES)
+
+                        # Comments must end with the proper punctuation.
+                        last_char_is_invalid = previous_line[-1] not in (
+                            ALLOWED_TERMINATING_PUNCTUATIONS)
+                        if (last_char_is_invalid and not (
+                                excluded_phrase_is_present_at_end or
+                                excluded_phrase_at_beginning_of_prev_line)):
+                            self.add_message(
+                                'invalid-punctuation-used',
+                                line=previous_line_num)
+
+                    # Comments may include a lowercase character at the
+                    # beginning only if they start with version info or a data
+                    # type or a variable name e.g. "# next_line is of string
+                    # type." or "# v2 version does not have ExplorationStats
+                    # Model." or "# int. The file size, in bytes.".
+                    if len(line) > 1 and space_at_beginning_of_comment:
+
+                        # Check if variable name is used.
+                        underscore_is_present = '_' in line.split()[1]
+                        if underscore_is_present:
+                            previous_line_num = line_num
+                            previous_line = line
+                            continue
 
-            # Comments must start with a capital letter.
-            if not previous_line.startswith(b'#') and (
-                    re.search(br'^# [a-z][A-Za-z]*.*$', line)):
-                self.add_message(
-                    'no-capital-letter-at-beginning', line=line_num + 1)
+                        # Check if allowed prefix is used.
+                        allowed_prefix_is_present = any(
+                            line[2:].startswith(word) for word in (
+                                allowed_comment_prefixes))
+                        if allowed_prefix_is_present:
+                            previous_line_num = line_num
+                            previous_line = line
+                            continue
+
+                    # Comments must start with a capital letter.
+                    self._check_no_capital_letter_at_beginning(line, line_num)
```

Comment:
Move this to the top as it's validating the first part of the comments and it will make it readable!

Reply:
Can't move it above then it will check if comment for a capital letter. We have to keep it here so that other validations are done first and this check is done at last.

Reply:
> Can't move it above then it will check if comment for a capital letter.

I'm not sure what's the issue with this, I think I didn't get it, can you please rephrase this?

Reply:
No longer using it here.